client/packages/programs/src/JsonForms/components/Note.tsx

```diff
@@ -0,0 +1,172 @@
+import { ControlProps, rankWith, uiTypeIs } from '@jsonforms/core';
+import { withJsonFormsControlProps } from '@jsonforms/react';
+import {
+  DetailInputWithLabelRow,
+  useDebounceCallback,
+  useTranslation,
+  useFormatDateTime,
+} from '@openmsupply-client/common';
+import { Typography } from '@mui/material';
+import { FORM_LABEL_WIDTH } from '../common/styleConstants';
+import { z } from 'zod';
+import { useZodOptionsValidation } from '../common/useZodOptionsValidation';
+
+const Options = z
+  .object({
+    width: z.string().optional(),
+    /**
+     * How many rows should the textbox display initially (default: 5)
+     */
+    rows: z.number().optional(),
+  })
+  .strict()
+  .optional();
+
+type Options = z.infer<typeof Options>;
+
+export const noteTester = rankWith(5, uiTypeIs('Note'));
+
+const UIComponent = (props: ControlProps) => {
+  const { data, handleChange, path, errors, uischema, config, enabled } = props;
+  const [localText, setLocalText] = useState<string | undefined>(data?.text);
+  const [localAuthor, setLocalAuthor] = useState<string | undefined>(
+    data?.authorName ??
+      // TO-DO: Use full name for default once available in database
+      config.user?.name
+  );
+  // timestamp of the last key stroke
+  const [latestKey, setLatestKey] = useState<number>(0);
+  const { localisedDateTime } = useFormatDateTime();
+
+  const { errors: zErrors, options: schemaOptions } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+
+  const error = !!errors || !!zErrors;
+
+  const onChange = (
+    text: string | undefined,
+    authorName: string | undefined
+  ) => {
+    const authorId = config.user?.id;
+    const created = data?.created ?? new Date().toISOString();
+    handleChange(
+      path,
+      error
+        ? undefined
+        : {
+            text,
+            authorId,
+            authorName,
+            created,
+          }
+    );
+  };
+
+  // debounce avoid rerendering the form on every key stroke which becomes a performance issue
+  const onChangeText = useDebounceCallback(
+    (value: string) => value && onChange(value, localAuthor),
+    [path, error, localAuthor]
+  );
+
+  const onChangeAuthor = useDebounceCallback(
+    (value: string) => value && onChange(localText, value),
+    [path, error, localText]
+  );
+
+  const t = useTranslation('programs');
+
+  const helperText = zErrors ?? errors;
+
+  useEffect(() => {
+    // Using debounce, the actual data is set after 500ms after the last key
+    // stroke (localDataTime). If data is set from the outside, e.g. through a
+    // reset, we want to update our local data as well. To distinguish between
+    // debounced events and external data updates we only take data that comes
+    // in at least 500ms after the last key stroke, i.e. it must be set from the
+    // outside.
+    if (Date.now() > latestKey + 500) {
+      setLocalText(data?.text);
+    }
+  }, [data]);
```

Comment:
Note, I almost copied this code for something else as well, so probably good to refactor it out at some point...

Reply:
Yeah, I'll make a new hook for it, since it seems pretty universal.

Reply:
I needed one for the IdGenerator so I created one there already...